crates/staged-sync/src/builder.rs

```diff
@@ -0,0 +1,317 @@
+use crate::config::{BodiesConfig, ExecutionConfig, SenderRecoveryConfig, TotalDifficultyConfig};
+
+use super::config::HeadersConfig;
+use reth_consensus::BeaconConsensus;
+use reth_db::database::Database;
+use reth_downloaders::{
+    bodies::concurrent::ConcurrentDownloader,
+    headers::linear::{LinearDownloadBuilder, LinearDownloader},
+};
+use reth_interfaces::{
+    consensus::{Consensus, ForkchoiceState},
+    sync::{NoopSyncStateUpdate, SyncStateUpdater},
+};
+use reth_network::{FetchClient, NetworkHandle};
+use reth_primitives::{ChainSpec, ChainSpecBuilder};
+use reth_stages::{
+    metrics::HeaderMetrics,
+    stages::{
+        bodies::BodyStage, execution::ExecutionStage, hashing_account::AccountHashingStage,
+        hashing_storage::StorageHashingStage, headers::HeaderStage, merkle::MerkleStage,
+        sender_recovery::SenderRecoveryStage, total_difficulty::TotalDifficultyStage,
+    },
+    Pipeline, PipelineEvent,
+};
+
+use std::sync::Arc;
+use tokio::sync::mpsc::Sender;
+
+use eyre::Result;
+
+/// Rough API usage:
+///
+/// 1. To launch the pipeline
+/// RethBuilder::new(..).with_execution(..).online(..).with_headers(..).with_bodies(..).build()
+/// RethBuilder::new(..).with_execution(..).build()
+///
+/// OR
+///
+/// 2. To make it easy to run an individual stage:
+///
+/// let builder = RethBuilder::new(..).online(..).with_headers(..);
+/// let header = builder.headers_stage().await?.expect("should be configured");
+/// let tx = Transaction::new(builder.db());
+/// let input = ExecInput {
+///     previous_stage: Some((StageId("No Previous Stage"), self.to)),
+///     stage_progress: Some(self.from),
+/// };
+/// stage.execute(&mut tx, input).await?;
+#[must_use = "need to call `build` on this struct"]
+#[derive(Debug, Default)]
+pub struct RethBuilder {
```

Comment:
I feel like this should be called Pipeline builder?

Reply:
We already have one